File: dijkstra/javascript/graph.js
'use strict';

module.exports = Graph;

function Graph(info) {
    this.vertexCount = info.vertexCount;
    this.edges = constructEdges(info);
}

function constructEdges(info) {
    var edges = initEdges(info.vertexCount);
    info.edges.forEach(function(edge) {
        insertEdge(edges, edge.x, edge.y, edge.distance, false);
    });
    return edges;
}

function initEdges(vertexCount) {
    var edges = new Array(vertexCount);
    for (var i = 0; i < vertexCount; i++) {
        edges[i] = [];
    }
    return edges;
}

function insertEdge(edges, x, y, distance, directed) {
    edges[x].push({y: y, distance: distance});
    if (!directed) {
        insertEdge(edges, y, x, distance, true);
    }
}

File: vertex_cover/rust/graph.rs
#[derive(Clone)]
pub struct Edge {
    pub x: usize,
    pub y: usize,
}

pub struct Graph {
    pub vertex_count: usize,
    pub edges: Vec<Edge>,
}

impl Edge {
    pub fn shares_vertex(&self, other: &Edge) -> bool {
        self.x == other.x || self.x == other.y || self.y == other.x || self.y == other.y
    }
}
